### ISSUE TITLE
vendoring member list makes serf useless as a library.

### ISSUE THREAD

**Issue by james-lawrence** [hashicorp/serf#386]
take for the following code example:
```go
import (
       "log"

	"github.com/hashicorp/serf/serf"
	"github.com/hashicorp/serf/vendor/github.com/hashicorp/memberlist"
)
type filter struct{}

func (t filter) NotifyAlive(peer *memberlist.Node) error {
	log.Println(peer.Name, peer.Addr.String(), int(peer.Port))
	return nil
}

func NewClusterOptions(nodeID, addr string, port int) Options {
	events := make(chan serf.Event)
	config := serf.DefaultConfig()
	config.MemberlistConfig = memberlist.DefaultWANConfig()
	config.MemberlistConfig.Alive = filter{}
	config.MemberlistConfig.BindAddr = addr
	config.MemberlistConfig.BindPort = port

	if nodeID != "" {
		config.NodeName = nodeID
	}

	return Options{
		Config:  config,
		EventCh: events,
	}
}

// Options holds the options for the cluster.
type Options struct {
	*serf.Config
	EventCh chan serf.Event
}
```

is code fails to compile because you can't import a packages vendored package. but if you try to import memberlist as a normal package:
```go
import (
"github.com/hashicorp/memberlist"
)
```
go complains about mismatched types.

**Comment by sean-** [comment]
This is earily similar to the long conversation happening on go-nuts:


>A tricky (impossible?) situation with common packages and vendoring - Google Groups

https://groups.google.com/forum/m/#!topic/golang-nuts/AnMr9NL6dtc

Thanks for bringing this to our attention. We'll get something figured out. Do you have this problem if you synchronize memberlist in vendor and your GOPATH?

**Comment by james-lawrence** [comment]
>Do you have this problem if you synchronize memberlist in vendor and your GOPATH?

yes because its based on the path, removing memberlist from my path also does not resolve the issue as memberlist still can't be located.

one solution is to move memberlist to a vendored directory closer to where its used, i.e.) the commands package. once that is done the problem goes away. Basically any exposed types can't be vendered in serf.

**Comment by slackpad** [comment]
@james-lawrence I updated memberlist to HEAD in https://github.com/hashicorp/serf/pull/387 - if you try the latest Serf now does that help your issue?

**Comment by slackpad** [comment]
Sorry that still won't fix it - just ran a local test from @sean-. Sorry about that.

**Comment by cbednarski** [comment]
@james-lawrence For now you can work around this problem using `GO15VENDOREXPERIMENT=0`. I believe this is a compiler bug in the new Go vendor behavior. Since the fallback logic `GO15...=0` is slated to be removed in 1.7 we will need upstream to fix this before that release.

**Comment by freeformz** [comment]
Remove or promote serf's vendor folder out of the serf library into your projects. I know this sucks though, but that's really the only way to handle this. Various tools (govendor, glide, gb ... and to a lesser extent because of time godep) are trying to make this easier.

**Comment by james-lawrence** [comment]
I don't actually need a fix personally, I was just pointing it out when I was playing with serf. Decided memberlist by itself will likely meet my needs.

**Comment by james-lawrence** [comment]
any progress on a solution for this? go1.7 is coming out soon making the current workaround listed here useless going forward.

**Comment by sean-** [comment]
@james-lawrence Give this a twirl and let us know.

**Comment by F21** [comment]
I am experiencing the same problem using glide to import serf into my project. Glide recommends importing at the base: `github.com/hashicorp/serf`.

Unfortunately, this means that the `vendor` folder is also downloaded, which means if I want to pass my own `memberlist.config` into `serf.config`, go complains about unmatched types:

```
cannot use memberlistConfig (type *"github.com/F21/myproject/vendor/github.com/hashicorp/memberlist".Config) as type *"github.com/F21/myproject/vendor/github.com/hashicorp/serf/vendor/github.com/hashicorp/memberlist".Config in assignment
```
Since there's already a `vendor/vendor.json`, I assume `govendor` is being used for this project. In that case, would it make sense to delete all the dependencies from `vendor/` and just keep `vendor.json`, so that serf is usable as a library?

**Comment by james-lawrence** [comment]
@sean- I think there was a misunderstanding of my suggested fix. It wasn't about moving the serf to cmd/serf. it was to move the vendor directory to be under the serf (now cmd/serf) directory.

**Comment by sean-** [comment]
I haven't looked at this in a while, but I do remember trying that and it failed to build or pass one of our tests and being a blocker for moving forward.  I believe the current situation was "as good as it was going to get" for the time being.  And please don't misunderstand, we're not thrilled with the current situation either, but it's better than where it was.

**Comment by james-lawrence** [comment]
@sean- no worries. I was clarifying my original thoughts. Not critizing the current situation. If I every have a spare moment on a weekend or for some reason I need to use serf as a library I'll take a crack at fixing it but thats unlikely given the current state of affairs.

**Comment by migueleliasweb** [comment]
I'm not as "fluent" as I wanted to be on Go but I wonder how hashicorp uses serf internally on consul. Is it bundled ? Like copy and paste ?

**Comment by sean-** [comment]
@migueleliasweb It's used as a library and hooked into the main Consul Agent routines.  If you dig around in https://github.com/hashicorp/consul/tree/master/command/agent you'll see that we listen for and respond to Serf events.

**Comment by victorcoder** [comment]
I'm experiencing the same issue using serf as a library.

**Comment by cbednarski** [comment]
@victorcoder sorry for the headache here. You will need to strip serf's vendor library when you pull it into you project.

As freeformz mentioned many go dependency management tools do this for you. If you are not using one you will need to delete serf's vendor folder from your GOPATH by hand.

Unfortunately there isn't a way to make this work for both library and cli use-cases out of the box with go's standard tooling.
